pkg/strategy/grid2/recover.go

```diff
@@ -0,0 +1,305 @@
+
+	expectedOrderNums := s.GridNum - 1
+	openOrdersOnGridNums := int64(len(openOrdersOnGrid))
+	s.debugLog("open orders nums: %d, expected nums: %d", openOrdersOnGridNums, expectedOrderNums)
+	if expectedOrderNums == openOrdersOnGridNums {
+		// no need to recover
+		return nil
+	} else if expectedOrderNums < openOrdersOnGridNums {
+		return fmt.Errorf("amount of grid's open orders should not > amount of expected grid's orders")
+	}
+
+	// 1. build pin-order map
+	pinOrdersOpen, err := s.buildPinOrderMap(grid.Pins, openOrdersOnGrid)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build pin order map with open orders")
+	}
+
+	// 2. build the filled pin-order map by querying trades
+	pinOrdersFilled, err := s.buildFilledPinOrderMapFromTrades(ctx, historyService, pinOrdersOpen)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build filled pin order map")
+	}
+
+	// 3. get the filled orders from pin-order map
+	filledOrders := pinOrdersFilled.AscendingOrders()
+	numFilledOrders := len(filledOrders)
+	if numFilledOrders == int(expectedOrderNums-openOrdersOnGridNums) {
+		// nums of filled order is the same as Size - 1 - num(open orders)
+	} else if numFilledOrders == int(expectedOrderNums-openOrdersOnGridNums+1) {
+		filledOrders = filledOrders[1:]
+	} else {
+		return fmt.Errorf("not reasonable num of filled orders")
+	}
+
+	// 4. verify the grid
+	if err := s.verifyFilledGrid(s.grid.Pins, pinOrdersOpen, filledOrders); err != nil {
+		return errors.Wrapf(err, "verify grid with error")
+	}
+
+	s.debugOrders("emit filled orders", filledOrders)
+
+	// 5. emit the filled orders
+	activeOrderBook := s.orderExecutor.ActiveMakerOrders()
+	for _, filledOrder := range filledOrders {
+		activeOrderBook.EmitFilled(filledOrder)
+	}
+
+	// 6. emit grid ready
+	s.EmitGridReady()
+
+	// 7. debug and send metrics
+	// wait for the reverse order to be placed
+	time.Sleep(2 * time.Second)
+	debugGrid(s.logger, grid, s.orderExecutor.ActiveMakerOrders())
+	s.updateGridNumOfOrdersMetricsWithLock()
+	s.updateOpenOrderPricesMetrics(s.orderExecutor.ActiveMakerOrders().Orders())
+
+	return nil
+}
+
+func (s *Strategy) verifyFilledGrid(pins []Pin, pinOrders PinOrderMap, filledOrders []types.Order) error {
+	s.debugLog("pins: %+v", pins)
+	s.debugLog("open pin orders:\n%s", pinOrders.String())
+	s.debugOrders("filled orders", filledOrders)
+
+	for _, filledOrder := range filledOrders {
+		price := filledOrder.Price
+		if o, exist := pinOrders[price]; !exist {
+			return fmt.Errorf("the price (%+v) is not in pins", price)
+		} else if o.OrderID != 0 {
+			return fmt.Errorf("there is already an order at this price (%+v)", price)
+		} else {
+			pinOrders[price] = filledOrder
+		}
+	}
+
+	s.debugLog("filled pin orders:\n%+v", pinOrders.String())
+
+	side := types.SideTypeBuy
+	for _, pin := range pins {
+		order, exist := pinOrders[fixedpoint.Value(pin)]
+		if !exist {
+			return fmt.Errorf("there is no order at price (%+v)", pin)
```

Comment:
use %s and pin.String() ?

Reply:
there is no String method for Pin. we need to convert it to fixedpoint.Value and then use string.
However I also want to remain the original value not after formating.